content/courses/Ada_For_The_C_Embedded_Developer/chapters/04_Embedded.rst

```diff
@@ -240,9 +240,194 @@ Dealing with Absence of FPU with Fixed Point
 Volatile and Atomic data
 ------------------------
 
-.. todo::
+Ada has built-in support for handling both volatile and atomic data. Let's
+start by discussing volatile objects.
+
+Volatile
+~~~~~~~~
+
+A `volatile <https://en.wikipedia.org/wiki/Volatile_(computer_programming)>`_
+can be described as an object in memory whose value may change between two
+consecutive memory accesses of a process A |mdash| even if process A itself
+hasn't changed the value. This situation may arise when an object in memory is
+being shared by multiple threads. For example, a thread *B* may modify the
+value of that object between two read accesses of a thread *A*. Another typical
+example is the one of
+`memory-mapped I/O <https://en.wikipedia.org/wiki/Memory-mapped_I/O>`_, where
+the hardware might be constantly changing the value of an object in memory.
+
+Because the value of a volatile object may be constantly changing, a compiler
+cannot store the value of that object into a register and use the value from
+the register in subsequent operations |mdash| because the value stored in the
+register would be outdated if another process had changed the volatile object
+in the meantime. Instead, the compiler must read the value of that object from
+memory for each access. In Ada, the standard specifies that, "for a volatile
```

Comment:
Unfortunately (or fortunately, depending on who you ask), the Ada RM no longer says that. The last version that did was Ada 2005. The wording is a good bit more subtle now, partly for the sake of multiprocessors. This is a very complicated topic...